### ISSUE TITLE
Lock sub-dependencies 

### ISSUE THREAD

**Issue by barth12** [yarnpkg/yarn#6840]
When reading yarn docs (the part: Selective dependency resolutions) I found usage case:

Your dependency defines a broad version range and your sub-dependency just got a problematic update so you want to pin it to an earlier version.

So there it is: let's say I installed dependency, and this dependency has its sub-dependencies which are defined by dependency lockfile. When the dependency has defined broad version of its sub-dependency the issue can occur, ie. sub-dependency can be bumped and it can be destructive to its mother dependency (then the whole app goes down) on the fresh install. This means that I have no control over sub-dependency version. On every fresh yarn install it can be very different even though I have my own lock file.

Please help me understand, first of all, even though I have my own lockfile I cannot control sub-dependencies versioning? second, is there a way to control it somehow in order to prevent uncontrolled sub-dependency bumping? Selective dependency resolutions is a way to control it but only after finding that error occurs and setting resolutions for ALL sub-dependencies makes no sense.

Expected behavior: lock sub-dependency to flat versions as it is with first-level dependencies so when fresh install occurs it will install not only exact dependency but also the same sub-dependency on other machine (version control).

Thank you!

**Comment by arcanis** [comment]
> On every fresh yarn install it can be very different even though I have my own lock file.

No. Your lockfile contains ALL the version used by your application, including transitive dependencies.

The selective resolution is for when you explicitly want to use a specific version above anything else, even if it wouldn't satisfy the range otherwise (for example a dependency is `^1.0.0`, but you want to use the release `2.0.0`).

**Comment by barth12** [comment]
@arcanis thank you for the answer. Of course, you're right. Still though it is easy to imagine a situation where sub-dependencies bumping occurs (within the range) but it crashes the dependency and hence the whole app. It is easy to manipulate semver of dependencies in order to flatten them and avoid the risk of unwanted bumping but it is much harder to do the same with sub-dependencies. Btw, does setting save prefix has an impact on sub-dependecies as well (I thought about flattening sub-dependencies versioning that way)?

**Comment by arcanis** [comment]
> Still though it is easy to imagine a situation where sub-dependencies bumping occurs (within the range) but it crashes the dependency and hence the whole app.

Ideally that's what semver is for - BC-breaking changes should only happen between major versions, so any accidental upgrade should be safe. Of course sometimes it isn't, but that's the exception more than the rule. Overall, Yarn tries not to upgrade things unless it really has to.

> Btw, does setting save prefix has an impact on sub-dependecies as well (I thought about flattening sub-dependencies versioning that way)?

No, that would imply changing the constraints set by the packages and Yarn doesn't do that.
